File: Common/mkfb.c
/*
 * Copyright (c) 1995-2024 Paul Mattes.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *     * Neither the names of Paul Mattes nor the names of his contributors
 *       may be used to endorse or promote products derived from this software
 *       without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY PAUL MATTES "AS IS" AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
 * EVENT SHALL PAUL MATTES BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
 * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
 * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/*
 *	mkfb.c
 *		Utility to create RDB string definitions from a simple
 *		#ifdef'd .ad file.
 *
 * 	mkfb [-c] [-o outfile] [infile...]
 */

#include <stdio.h>
#include <stdbool.h>
#include "conf.h"
#if defined(_WIN32) /*[*/
# include "wincmn.h"
#else /*][*/
# include <string.h>
# include <ctype.h>
# include <stdlib.h>
# include <errno.h>
#endif /*]*/

#define BUFSZ	1024		/* input line buffer size */
#define ARRSZ	8192		/* output array size */
#define SSSZ	10		/* maximum nested ifdef */

unsigned aix[ARRSZ];		/* fallback array indices */
unsigned xlno[ARRSZ];		/* fallback array line numbers */
unsigned n_fallbacks = 0;	/* number of fallback entries */

/* ifdef state stack */
#define MODE_COLOR	0x00000001
#define MODE_APL	0x00000002
#define MODE_DBCS	0x00000004
#define MODE__WIN32	0x00000008

#define MODEMASK	0x00000fff

struct {
    unsigned long ifdefs;
    unsigned long ifndefs;
    unsigned lno;
} ss[SSSZ];
unsigned ssp = 0;

struct {
    const char *name;
    unsigned long mask;
} parts[] = {
    { "COLOR", MODE_COLOR },
    { "X3270_APL", MODE_APL },
    { "X3270_DBCS", MODE_DBCS },
    { "_WIN32", MODE__WIN32 }
};
#define NPARTS	(sizeof(parts)/sizeof(parts[0]))

unsigned long is_defined =
    MODE_COLOR | MODE_APL | MODE_DBCS;
unsigned long is_undefined;

char *me;

void emit(FILE *t, int ix, char c);

void
usage(void)
{
    fprintf(stderr, "usage: %s [-c] -[w] [-o outfile] [infile...]\n",
	    me);
    exit(1);
}

/*
* Wrapper around Windows' brain-dead tmpfile().
*/
static FILE *
mkfb_tmpfile(void)
{
    FILE *f;
#if defined(_WIN32) /*[*/
    char *n;
#endif /*]*/

#if !defined(_WIN32) /*[*/
    f = tmpfile();
    if (f == NULL) {
	perror("tmpfile");
	exit(1);
    }
#else /*][*/
    n = _tempnam(NULL, "mkfb");
    if (n == NULL) {
	fprintf(stderr, "_tempnam failed.\n");
	exit(1);
    }
    f = fopen(n, "w+b");
    if (f == NULL) {
	fprintf(stderr, "_tempnam open(\"%s\") failed: %s\n", n,
		strerror(errno));
	exit(1);
    }
    free(n);
#endif /*]*/

    return f;
}

int
main(int argc, char *argv[])
{
    int ix;
    char *ofile = NULL;
    char buf[BUFSZ];
    int lno = 0;
    int cc = 0;
    unsigned i;
    bool continued = false;
    const char *filename = "standard input";
    FILE *u, *t, *tc = NULL, *tm = NULL;
    int cmode = 0;
    unsigned long ifdefs;
    unsigned long ifndefs;
    bool last_continue = false;
    int infiles = 0;

    /* Parse arguments. */
    if ((me = strrchr(argv[0], '/')) != NULL) {
	me++;
    } else {
	me = argv[0];
    }
    for (ix = 1; ix < argc; ix++) {
	if (argv[ix][0] != '-') {
	    break;
	}
	if (!strcmp(argv[ix], "-c")) {
	    cmode = 1;
	} else if (!strcmp(argv[ix], "-w")) {
	    is_defined |= MODE__WIN32;
	} else if (!strcmp(argv[ix], "-o")) {
	    if (argc < ix + 1) {
		usage();
	    }
	    ix++;
	    ofile = argv[ix];
	} else {
	    usage();
	}
    }
    infiles = argc - ix;

    /* Open the output file. */
    if (ofile) {
	if (freopen(ofile, "w", stdout) == NULL) {
	    perror(ofile);
	    exit(1);
	}
    }

    is_undefined = MODE_COLOR | (~is_defined & MODEMASK);

    /* Start a temp file. */
    u = mkfb_tmpfile();

    /* Do #ifdef, comment and whitespace processing first. */
    while (!infiles || ix < argc) {

	if (infiles) {
	    if (freopen(argv[ix], "r", stdin) == NULL) {
		perror(argv[ix]);
		exit(1);
	    }
	    filename = argv[ix];
	    ix++;
	}

	while (fgets(buf, BUFSZ, stdin) != NULL) {
	    char *s = buf;
	    size_t sl;
	    unsigned i;

	    lno++;

	    /* Skip leading white space. */
	    while (isspace((unsigned char)*s)) {
		s++;
	    }
	    if (cmode &&
		(!strncmp(s, "x3270.", 6) || !strncmp(s, "x3270*", 6))) {
		s += 6;
	    }

	    /* Remove trailing white space. */
	    while ((sl = strlen(s)) && isspace((unsigned char)s[sl-1])) {
		s[sl-1] = '\0';
	    }

	    /* Skip comments and empty lines. */
	    if ((!last_continue && *s == '!') || !*s) {
		continue;
	    }

	    /* Check for simple if[n]defs. */
	    if (*s == '#') {
		int ifnd = 1;

		if (!strncmp(s, "#ifdef ", 7) ||
			!(ifnd = strncmp(s, "#ifndef ", 8))) {
		    char *tk;

		    if (ssp >= SSSZ) {
			fprintf(stderr, "%s, line %d: Stack overflow\n",
				filename, lno);
			exit(1);
		    }
		    ss[ssp].ifdefs = 0L;
		    ss[ssp].ifndefs = 0L;
		    ss[ssp].lno = lno;

		    tk = s + 7 + !ifnd;
		    for (i = 0; i < NPARTS; i++) {
			if (!strcmp(tk, parts[i].name)) {
			    if (!ifnd) {
				ss[ssp++].ifndefs = parts[i].mask;
			    } else {
				ss[ssp++].ifdefs = parts[i].mask;
			    }
			    break;
			}
		    }
		    if (i >= NPARTS) {
			fprintf(stderr, "%s, line %d: Unknown condition\n",
				filename, lno);
			exit(1);
		    }
		    continue;
		} else if (!strcmp(s, "#else")) {
		    unsigned long tmp;

		    if (!ssp) {
			fprintf(stderr, "%s, line %d: Missing #if[n]def\n",
				filename, lno);
			exit(1);
		    }
		    tmp = ss[ssp-1].ifdefs;
		    ss[ssp-1].ifdefs = ss[ssp-1].ifndefs;
		    ss[ssp-1].ifndefs = tmp;
		} else if (!strcmp(s, "#endif")) {
		    if (!ssp) {
			fprintf(stderr, "%s, line %d: Missing #if[n]def\n",
				filename, lno);
			exit(1);
		    }
		    ssp--;
		} else {
		    fprintf(stderr, "%s, line %d: Unrecognized # directive\n",
			    filename, lno);
		    exit(1);
		}
		continue;
	    }

	    /* Figure out if there's anything to emit. */

	    /* First, look for contradictions. */
	    ifdefs = 0;
	    ifndefs = 0;
	    for (i = 0; i < ssp; i++) {
		ifdefs |= ss[i].ifdefs;
		ifndefs |= ss[i].ifndefs;
	    }
	    if (ifdefs & ifndefs) {
#ifdef DEBUG_IFDEFS
		fprintf(stderr, "contradiction, line %d\n", lno);
#endif
		continue;
	    }

	    /* Then, apply the actual values. */
	    if (ifdefs && (ifdefs & is_defined) != ifdefs) {
#ifdef DEBUG_IFDEFS
		fprintf(stderr, "ifdef failed, line %d\n", lno);
#endif
		continue;
	    }
	    if (ifndefs && (ifndefs & is_undefined) != ifndefs) {
#ifdef DEBUG_IFDEFS
		fprintf(stderr, "ifndef failed, line %d\n", lno);
#endif
		continue;
	    }

	    /* Emit the text. */
	    fprintf(u, "%lx %lx %d\n%s\n", ifdefs, ifndefs, lno, s);
	    last_continue = strlen(s) > 0 && s[strlen(s) - 1] == '\\';
	}

	if (!infiles) {
	    break;
	}
    }

    if (ssp) {
	fprintf(stderr, "%d missing #endif(s)\n", ssp);
	fprintf(stderr, "last #ifdef was at line %u\n", ss[ssp-1].lno);
	exit(1);
    }

    /* Re-scan, emitting code this time. */
    rewind(u);
    t = mkfb_tmpfile();
    if (!cmode) {
	tc = mkfb_tmpfile();
	tm = mkfb_tmpfile();
    }

    /* Emit the initial boilerplate. */
    fprintf(t, "/* This file was created automatically by mkfb. */\n\n");
    fprintf(t, "#include \"globals.h\"\n");
    fprintf(t, "#include \"fallbacks.h\"\n");
    if (cmode) {
	fprintf(t, "static unsigned char fsd[] = {\n");
    } else {
	fprintf(t, "unsigned char common_fallbacks[] = {\n");
	fprintf(tc, "unsigned char color_fallbacks[] = {\n");
	fprintf(tm, "unsigned char mono_fallbacks[] = {\n");
    }

    /* Scan the file, emitting the fsd array and creating the indices. */
    while (fscanf(u, "%lx %lx %d\n", &ifdefs, &ifndefs, &lno) == 3) {
	char *s = buf;
	char c;
	int white;
	FILE *t_this = t;
	int ix = 0;

	if (fgets(buf, BUFSZ, u) == NULL) {
	    break;
	}
	if (strlen(buf) > 0 && buf[strlen(buf)-1] == '\n') {
	    buf[strlen(buf)-1] = '\0';
	}

#if 0
	fprintf(stderr, "%lx %lx %d %s\n", ifdefs, ifndefs, lno, buf);
#endif

	/* Add array offsets. */
	if (cmode) {
	    /* Ignore color.  Accumulate offsets into an array. */
	    if (n_fallbacks >= ARRSZ) {
		fprintf(stderr, "Buffer overflow\n");
		exit(1);
	    }
	    if (!continued) {
		aix[n_fallbacks] = cc;
		xlno[n_fallbacks++] = lno;
	    }
	} else {
	    /* Use color to decide which file to write into. */
	    if (!(ifdefs & MODE_COLOR) && !(ifndefs & MODE_COLOR)) {
		/* Both. */
		t_this = t;
		ix = 0;
	    } else if (ifdefs & MODE_COLOR) {
		/* Just color. */
		t_this = tc;
		ix = 1;
	    } else {
		/* Just mono. */
		t_this = tm;
		ix = 2;
	    }
	}

	continued = false;
	white = 0;
	while ((c = *s++) != '\0') {
	    if (c == ' ' || c == '\t') {
		white++;
	    } else if (white) {
		emit(t_this, ix, ' ');
		cc++;
		white = 0;
	    }
	    switch (c) {
	    case ' ':
	    case '\t':
		break;
	    case '#':
		if (!cmode) {
		    emit(t_this, ix, '\\');
		    emit(t_this, ix, '#');
		    cc += 2;
		} else {
		    emit(t_this, ix, c);
		    cc++;
		}
		break;
	    case '\\':
		if (*s == '\0') {
		    continued = true;
		    break;
		} else if (cmode) {
		    switch ((c = *s++)) {
		    case 't':
			c = '\t';
			break;
		    case 'n':
			c = '\n';
			break;
		    default:
			break;
		    }
		}
		/* else fall through */
	    default:
		emit(t_this, ix, c);
		cc++;
		break;
	    }
	}
	if (white) {
	    emit(t_this, ix, ' ');
	    cc++;
	    white = 0;
	}
	if (!continued) {
	    if (cmode) {
		emit(t_this, ix, 0);
	    } else {
		emit(t_this, ix, '\n');
	    }
	    cc++;
	}
    }
    fclose(u);
    if (cmode) {
	fprintf(t, "};\n\n");
    } else {
	emit(t, 0, 0);
	fprintf(t, "};\n\n");
	emit(tc, 0, 0);
	fprintf(tc, "};\n\n");
	emit(tm, 0, 0);
	fprintf(tm, "};\n\n");
    }

    /* Copy tmp to output. */
    rewind(t);
    if (!cmode) {
	rewind(tc);
	rewind(tm);
    }
    while (fgets(buf, sizeof(buf), t) != NULL) {
	printf("%s", buf);
    }
    if (!cmode) {
	while (fgets(buf, sizeof(buf), tc) != NULL) {
	    printf("%s", buf);
	}
	while (fgets(buf, sizeof(buf), tm) != NULL) {
	    printf("%s", buf);
	}
    }

    if (cmode) {
	/* Emit the fallback array. */
	printf("char *fallbacks[%u] = {\n", n_fallbacks + 1);
	for (i = 0; i < n_fallbacks; i++) {
	    printf("\t(char *)&fsd[%u], /* line %u */\n", aix[i],
		    xlno[i]);
	}
	printf("\tNULL\n};\n\n");
    }

    /* Emit some test code. */
    printf("%s", "#if defined(DEBUG) /*[*/\n\
#include <stdio.h>\n\
int\n\
main(int argc, char *argv[])\n\
{\n\
    int i;\n\
\n");
    if (cmode) {
	printf("%s",
"    for (i = 0; fallbacks[i] != NULL; i++) {\n\
	printf(\"%d: %s\\n\", i, fallbacks[i]);\n\
    }\n");
    } else {
	printf("%s",
"    printf(\"Common:\\n%s\\n\", common_fallbacks);\n\
    printf(\"Color:\\n%s\\n\", color_fallbacks);\n\
    printf(\"Mono:\\n%s\\n\", mono_fallbacks);\n");
    }
    printf("%s",
"    return 0;\n\
}\n");
    printf("#endif /*]*/\n");

    fflush(stdout);
    fclose(stdout);
    fclose(t);
    if (!cmode) {
	fclose(tc);
	fclose(tm);
    }

    return 0;
}

static int n_out[3] = { 0, 0, 0 };

void
emit(FILE *t, int ix, char c)
{
    if (n_out[ix] >= 19) {
	fprintf(t, "\n");
	n_out[ix] = 0;
    }
    fprintf(t, "%3d,", (unsigned char)c);
    n_out[ix]++;
}

File: Common/mkfb.py
#!/usr/bin/env python3
#
# Copyright (c) 2025 Paul Mattes.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#     * Redistributions of source code must retain the above copyright
#       notice, this list of conditions and the following disclaimer.
#     * Redistributions in binary form must reproduce the above copyright
#       notice, this list of conditions and the following disclaimer in the
#       documentation and/or other materials provided with the distribution.
#     * Neither the names of Paul Mattes nor the names of his contributors
#       may be used to endorse or promote products derived from this software
#       without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY PAUL MATTES "AS IS" AND ANY EXPRESS OR IMPLIED
# WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
# EVENT SHALL PAUL MATTES BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
# OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
# WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
# OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
# ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
# Utility to create RDB string definitions from a simple #ifdef'd .ad file.
#  mkfb [-c] [-o outfile] [infile...]

from enum import Flag
import os
import sys

# #ifdef bit masks.
class Mode(Flag):
    NONE = 0
    COLOR = 0x1
    APL = 0x2
    X3270_DBCS = 0x4
    _WIN32 = 0x8
    MASK = 0xfff

# Globals.
cmode = False
is_defined = Mode.COLOR | Mode.APL | Mode.X3270_DBCS
is_undefined = Mode.NONE

class Ss:
    '''#ifdef stack element'''
    def __init__(self):
        self.ifdefs = Mode.NONE
        self.ifndefs = Mode.NONE
        self.lno = 0

# Set up ss (the #ifdef nesting stack) as an array for now.
# Clearly this should be something more Pythonic.
SSSZ = 10
ss = [Ss() for i in range(SSSZ)]
ssp = 0

class FilteredRecord:
    '''#ifdef temp'''
    def __init__(self, ifdefs: Mode, ifndefs: Mode, lno: int, filename: str, value: str):
        self.ifdefs = ifdefs
        self.ifndefs = ifndefs
        self.lno = lno
        self.filename = os.path.basename(filename)
        self.value = value

# Dictionary mapping #ifdef symbols to bit masks.
parts = {
    Mode.COLOR.name: Mode.COLOR,
    Mode.APL.name: Mode.APL,
    Mode.X3270_DBCS.name: Mode.X3270_DBCS,
    Mode._WIN32.name: Mode._WIN32
}

class OutBuffer:
    n_per_line = 19
    def __init__(self):
        self.text = []
        self.n_out = 0
    def append(self, text: str):
        '''Append a line of text'''
        self.text.append(text)
    def emit(self, c:str):
        '''Emit a byte, with nice formatting'''
        if self.n_out >= OutBuffer.n_per_line:
            self.append('')
            self.n_out = 0
        self.text[-1] += f'{ord(c):3d},'
        self.n_out += 1

def mkfb(infiles, ofile: str):
    '''Generate the fallback files'''

    last_continue = False
    lno = 0
    ssp = 0

    is_undefined = Mode.COLOR | (~is_defined & Mode.MASK)

    # Start accumulating filtered output.
    filtered = []

    # Do #ifdef, comment and whitespace processing first.
    while True:
        if infiles != None:
            if infiles == []:
                break
            in_file = open(infiles[0])
            filename = infiles[0]
            infiles = infiles[1:]
            lno = 0
        else:
            in_file = sys.stdin
            filename = 'standard input'

        # Process.
        while True:
            buf = in_file.readline()
            if buf == '':
                break
            lno += 1

            s = buf.strip()
            if cmode and (s.startswith('x3270.') or s.startswith('x3270*')):
                s = s[6:]

            if (not last_continue and s.startswith('!')) or len(s) == 0:
                continue

            if s.startswith('#'):
                is_ifndef = s.startswith('#ifndef')
                if s.startswith('#ifdef') or is_ifndef:
                    if ssp >= SSSZ:
                        print(f'{filename}:{lno}: Stack overflow', file=sys.stderr)
                        exit(1)
                    ss[ssp].ifdefs = Mode.NONE
                    ss[ssp].ifndefs = Mode.NONE
                    ss[ssp].lno = lno

                    tk = 6 + (1 if is_ifndef else 0)
                    condition = s[tk:].lstrip()
                    if condition in parts:
                        if is_ifndef:
                            ss[ssp].ifndefs = parts[condition]
                        else:
                            ss[ssp].ifdefs = parts[condition]
                        ssp += 1
                    else:
                        print(f'{filename}:{lno}: Unknown condition {condition}', file=sys.stderr)
                        exit(1)
                    continue
                elif s == '#else':
                    if ssp == 0:
                        print(f'{filename}:{lno}: Missing #if[n]def', file=sys.stderr)
                        exit(1)
                    tmp = ss[ssp-1].ifdefs
                    ss[ssp-1].ifdefs = ss[ssp-1].ifndefs
                    ss[ssp-1].ifndefs = tmp
                elif s == '#endif':
                    if ssp == 0:
                        print(f'{filename}:{lno}: Missing #if[n]def', file=sys.stderr)
                        exit(1)
                    ssp -= 1
                else:
                    print(f'{filename}:{lno}: Unrecognized # directive', file=sys.stderr)
                    exit(1)
                continue

            # Figure out if there's anything to emit.
            # First, look for contradictions.
            ifdefs = Mode.NONE
            ifndefs = Mode.NONE
            for i in range(ssp):
                ifdefs |= ss[i].ifdefs
                ifndefs |= ss[i].ifndefs
            if ifdefs != Mode.NONE and ifndefs != Mode.NONE:
                continue

            # Then, apply the actual values.
            if ifdefs != Mode.NONE and (ifdefs & is_defined) != ifdefs:
                continue
            if ifndefs != Mode.NONE and (ifndefs & is_undefined) != ifndefs:
                continue

            # Emit the text.
            filtered.append(FilteredRecord(ifdefs, ifndefs, lno, filename, s))
            last_continue = s.endswith('\\')

        if infiles != None:
            in_file.close()
        else:
            break
    
    if ssp != 0:
        print(f'{ssp} missing #endifs', file=sys.stderr)
        print(f'last #ifdef was at line {ss[ssp-1].lno}', file=sys.stderr)
        exit(1)

    # Re-scan, emitting code this time.
    t = OutBuffer()
    if not cmode:
        tc = OutBuffer()
        tm = OutBuffer()
    
    # Emit the initial boilerplate.
    t.append('/* This file was created automatically by mkfb. */')
    t.append('')
    t.append('#include "globals.h"',)
    t.append('#include "fallbacks.h"')
    t.append('')
    if cmode:
        t.append('static unsigned char fsd[] = {')
        t.append('')
    else:
        t.append('unsigned char common_fallbacks[] = {')
        t.append('')
        tc.append('unsigned char color_fallbacks[] = {')
        tc.append('')
        tm.append('unsigned char mono_fallbacks[] = {')
        tm.append('')

    # Scan the file, emitting the fsd array and creating the indices.
    cc = 0
    aix = []
    xlno = []
    xfilename = []
    backslash = False
    for rec in filtered:
        ifdefs = rec.ifdefs
        ifndefs = rec.ifndefs
        lno = rec.lno
        buf = rec.value

        if cmode:
            # Ignore color. Accumulate offsets into an array.
            t_this = t
            if not backslash:
                aix.append(cc)
                xlno.append(lno)
                xfilename.append(rec.filename)
        else:
            # Use color to decide which list to write into.
            if not Mode.COLOR in ifdefs and not Mode.COLOR in ifndefs:
                # Both.
                t_this = t
            elif Mode.COLOR in ifdefs:
                # Just color.
                t_this = tc
            else:
                # Just mono.
                t_this = tm

        backslash = False
        white = 0
        for c in buf:
            if backslash:
                if cmode:
                    if c == 't':
                        c = '\t'
                    elif c == 'n':
                        c = '\n'
                else:
                    t_this.emit('\\')
                    cc += 1
                backslash = False
            if c == ' ' or c == '\t':
                white += 1
            elif white > 0:
                t_this.emit(' ')
                cc += 1
                white = 0
            if c != ' ' and c != '\t':
                if c == '#':
                    if not cmode:
                        t_this.emit('\\')
                        t_this.emit('#')
                        cc += 2
                    else:
                        t_this.emit(c)
                        cc += 1
                elif c == '\\':
                    backslash = True
                else:
                    t_this.emit(c)
                    cc += 1
        if white > 0:
            # Line ended with whitespace pending.
            t_this.emit(' ')
            cc += 1
            white = 0
        if not backslash:
            # Line is not a continuation, terminate it.
            if cmode:
                t_this.emit('\0')
            else:
                t_this.emit('\n')
            cc += 1

    if cmode:
        t.append('};',)
        t.append('')
    else:
        t.emit('\0')
        t.append('};',)
        t.append('')
        tc.emit('\0')
        tc.append('};')
        tc.append('')
        tm.emit('\0')
        tm.append('};')
        tm.append('')

    # Copy tmp to output.

    # Open the output file.
    if ofile != None:
        out_file = open(ofile, 'w')
    else:
        out_file = sys.stdout
    for line in t.text:
        out_file.write(line + '\n')
    if not cmode:
        for line in tc.text:
            out_file.write(line + '\n')
        for line in tm.text:
            out_file.write(line + '\n')
    
    if cmode:
        # Emit the fallback array.
        print(f'char *fallbacks[{len(aix)+1}] = ' + '{', file=out_file)
        for i in range(len(aix)):
            print(f'    (char *)&fsd[{aix[i]}], /* {xfilename[i]}:{xlno[i]} */', file=out_file)
        print('    NULL\n};', file=out_file)
        print(file=out_file)

    # Emit some test code.
    print('#if defined(DEBUG) /*[*/', file=out_file)
    print('#include <stdio.h>', file=out_file)
    print('int', file=out_file)
    print('main(int argc, char *argv[])', file=out_file)
    print('{', file=out_file)
    print('    int i;', file=out_file)
    print(file=out_file)
    if cmode:
        print('    for (i = 0; fallbacks[i] != NULL; i++) {', file=out_file)
        print('\tprintf("%d: %s\\n", i, fallbacks[i]);', file=out_file)
        print('    }', file=out_file)
    else:
        print('    printf(\"Common:\\n%s\\n\", common_fallbacks);', file=out_file)
        print('    printf(\"Color:\\n%s\\n\", color_fallbacks);', file=out_file)
        print('    printf(\"Mono:\\n%s\\n\", mono_fallbacks);', file=out_file)
    print('    return 0;', file=out_file)
    print('}', file=out_file)
    print('#endif /*]*/', file=out_file)

def usage():
    '''Print a usage message and exit'''
    print('usage: mkfb [-c] -[w] [-o outfile] [infile...]', file=sys.stderr)
    exit(1)

# Parse the command line.
args = sys.argv[1:].copy()
infiles = None
ofile = None
while len(args) > 0:
    arg = args[0]
    if not arg.startswith('-'):
        infiles = args
        break
    if arg == '-c':
        cmode = True
    elif arg == '-w':
        is_defined |= Mode._WIN32
    elif arg == '-o':
        if len(args) < 2:
            usage()
        args = args[1:]
        ofile = args[0]
    else:
        usage()
    args = args[1:]

# Run.
mkfb(infiles, ofile)